ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/keys/ContractKeysValidation.scala

```diff
@@ -72,25 +72,18 @@ private[transaction] object ContractKeysValidation {
   )(implicit loggingContext: LoggingContext): StepResult[DamlTransactionEntrySummary] = {
     type KeyValidationStackStatus = Either[KeyValidationError, List[KeyValidationState]]
 
-    val keysValidationOutcome = transactionEntry.transaction
-      .foldInExecutionOrder[KeyValidationStackStatus](
-        Right(List(KeyValidationState(activeStateKeys = contractKeyDamlStateKeys)))
-      )(
-        exerciseBegin = (status, _, exerciseBeginNode) => {
-          val newStatus =
-            onCurrentStatus(status)(
-              checkNodeContractKey(exerciseBeginNode, contractKeysToContractIds, _)
-            )
-          (newStatus, true)
-        },
-        rollbackBegin = (status, _, _) =>
-          // Store state to restore after rollback
-          (status.map(stack => stack.head +: stack), true),
-        leaf = (status, _, leafNode) =>
-          onCurrentStatus(status)(checkNodeContractKey(leafNode, contractKeysToContractIds, _)),
-        exerciseEnd = (accum, _, _) => accum,
-        rollbackEnd = (status, _, _) => status.map(stack => stack.tail),
-      )
+    val transaction = transactionEntry.transaction
+
+    val keysValidationOutcome = for {
+      keyInputs <- transaction.contractKeyInputs.left.map(_ => Duplicate)
+      _ <- keyInputs.toList.traverse_ { case (key, optCid) =>
```

Comment:
Didn’t really seem worth factoring this out given that this is the only thing this function does but happy to factor it out in some form if you prefer.